File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static StringBuilder sb = new StringBuilder();
    static List<Integer>[] graph;
    static boolean[] visited;
    static int N, M, V;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        V = Integer.parseInt(st.nextToken());

        graph = new ArrayList[N + 1];
        visited = new boolean[N + 1];

        for (int i = 1; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            StringTokenizer st1 = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st1.nextToken());
            int b = Integer.parseInt(st1.nextToken());
            graph[a].add(b);
            graph[b].add(a);
        }

        for (int i = 1; i <= N; i++) {
            Collections.sort(graph[i]);
        }

        dfs(V);
        sb.append("\n");

        visited = new boolean[N + 1];
        bfs(V);

        System.out.println(sb);
    }


    static void dfs(int start) {
        visited[start] = true;
        sb.append(start).append(" ");

        for (Integer i : graph[start]) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(i);
            }
        }
    }

    static void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        visited[start] = true;
        q.offer(start);

        while (!q.isEmpty()) {
            int cur = q.poll();
            sb.append(cur).append(" ");
            for (Integer i : graph[cur]) {
                if (!visited[i]) {
                    visited[i] = true;
                    q.offer(i);
                }
            }
        }

    }
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
# dfs bfs 탐색
# 정점 여러개 -> 작은것 부터
# n:정점, m:간선, v:시작번호

from collections import deque

def dfs(v, graph, visited):
    visited[v] = True
    answer_dfs.append(v)
    for w in graph[v]:
        if not visited[w]:
            dfs(w, graph, visited)


def bfs(v, graph, visited):
    visited[v] = True
    answer_bfs.append(v)
    q = deque([v])
    while q:
        l = q.popleft()
        for w in graph[l]:
            if not visited[w]:
                visited[w] = True
                answer_bfs.append(w)
                q.append(w)


n, m, v = map(int, input().split())
graph = [[] for _ in range(n + 1)]
answer_dfs = []
answer_bfs = []

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for g in graph:
    g.sort()

dfs(v, graph, [False] * (n+1))
bfs(v, graph, [False] * (n+1))

print(*answer_dfs)
print(*answer_bfs)